PPM-334 slave: handle_1905_higher_layer_data_message: move from backhaul to son_slave

orenvor opened the pull request:
Move handling of 1905_higher_layer_data messages from the backhual manager to the son_slave thread.

https://jira.prplfoundation.org/browse/PPM-334

adam1985d commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -40,6 +40,7 @@
 #include <tlvf/wfa_map/tlvChannelSelectionResponse.h>
 #include <tlvf/wfa_map/tlvClientAssociationControlRequest.h>
 #include <tlvf/wfa_map/tlvClientAssociationEvent.h>
+#include <tlvf/wfa_map/tlvHigherLayerData.h>
```
Please also remove this include from the `backhaul_manager.cpp` file if possible

orenvor replied:
include removed from `backhaul_manager.cpp`.

adam1985d commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2321,9 +2321,6 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE: {
         return handle_ap_capability_query(cmdu_rx, src_mac);
     }
```
the commit message should follow the `contribution.md` rules.

Please use this PR to ramp-up on them.
(just from a quick review: line-length limits of title and body are not met)

morantr replied:
Also, the commit message is not accurate.
It is not that the handler should be on son_slave. It should be on the future Unified Agent, which the son_slave will be converted to.

I noticed a typo: "...match the new **behiver**" -> "...match the new **behavior**".
I really recommend you, to install `Grammarly` extension to chrome, and past there the commit message before committing. We are not native English speakers, and this tool can save you a lot of rework.

Last thing. You don't need to supply full Jira link, `PPM-334` is enough.

orenvor replied:
fixed

adam1985d commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+bool slave_thread::handle_1905_higher_layer_data_message(Socket *sd,
+                                                         ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    //only one backhaul manager (the slave) should return ACK for higher layer data message.
```
```suggestion
    // Only one backhaul manager (the slave) should return ACK for higher layer data message.
```
or anything else you think can make it clearer.

In general - although it is porting existing code from point A to point B, you can always rework is where you find it suitable.
You can use there small PRs to ramp on the coding guidelines and enforce them to the ported code.

adam1985d commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+    const auto protocol       = tlvHigherLayerData->protocol();
+    const auto payload_length = tlvHigherLayerData->payload_length();
+    LOG(DEBUG) << "protocol: " << std::hex << int(protocol);
+    LOG(DEBUG) << "payload_length: " << std::hex << int(payload_length);
```
```suggestion
    LOG(DEBUG) << "Protocol: " << std::hex << int(protocol);
    LOG(DEBUG) << "Payload-Length: " << std::hex << int(payload_length);
```
we should print the meaning of the data printed instead of the parameter name. this also makes future naming changes (that do not change the meaning) not required log print changes.

that being said:
1. please verify if the "payload_length" print is expected by some certification test before you change the print to "Payload-Length" and "Protocol"
2. if you change this and the certification tests don't use this print as reference (only our test-flows) then you must align our test-flows to the new print

orenvor replied:
suggestion applied.

adam1985d commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+    LOG(DEBUG) << "protocol: " << std::hex << int(protocol);
+    LOG(DEBUG) << "payload_length: " << std::hex << int(payload_length);
+
+    // build ACK message CMDU
```
```suggestion
    // Build ACK message CMDU
```

orenvor replied:
fixed

adam1985d commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << int(mid);
```
```suggestion
    LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << int(mid);
```

orenvor replied:
fixed

adam1985d commented on tests/test_flows.py:
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
+                  start_line: int = 0, fail_when_not_found: bool = True) -> bool:
+        '''Verify that the logfile for "entity_or_radio" matches "regex",
+           fail if no match when fail_when_not_found is enabled.
+        '''
```
```suggestion
        '''Verify that the log-file for "entity_or_radio" matches "regex",
           fail if no match is found when fail_when_not_found is enabled.
        '''
```

orenvor replied:
fixed

adam1985d commented on tests/test_flows.py:
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
+        if (received_in_radio0 == received_in_radio1):
+            self.fail(f"higher layer data message was received and acknowledged by "
+                      f"{int(received_in_radio0) + int(received_in_radio1)} agent's radios, "
+                      f"expected only 1")
```
```suggestion
                      f"expected exactly 1")
```
1. Since you decided to include in the same condition:
false,false
and
true,true
then your comment should reflect that.

2. This is a test so I would consider the following comment a nitpick - but:
I prefer the condition and the error-condition to be pre-acquired to the point you use them.
Meaning:
- Calculate the number of agents that received the message
- Then, compare it against the expected value.
- Then, print the expected and actual as part of the error log print

orenvor replied:
suggestion applied

adam1985d commented on tests/test_flows.py:
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
-        self.check_log(env.agents[0], r"payload_length: 0x4b0")
+        debug("Confirming matching protocol and payload length")
+        self.check_log(received_agent_radio, r"protocol: 0")
+        self.check_log(received_agent_radio, r"payload_length: 0x4b0")
```
please verify if the "payload_length" print is expected by some certification test before you change the print to "Payload-Length" and "Protocol"

morantr replied:
Certification tests don't scan log files, only our internal `test_flows` script.

orenvor replied:
suggestion applied.

adam1985d commented on tests/test_flows.py:
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
+                      f"expected only 1")
 
-        debug("Confirming matching protocol and payload length")
+        received_agent_radio = (env.agents[0].radios[0] if received_in_radio0
+                                else env.agents[0].radios[1])
```
do we care for debug purposes on which radio we received the message?
if yes, then need to add debug print regarding this.
@morantr

morantr replied:
It is used for the next check on the log files. Once we know which son_slave received it, we don't need to check on both of them.

adam1985d reviewed (changes requested):
In general, I would have approved this PR (since the changes are small and don't require a re-review).
That being said - since these are your first PRs, I want to re-review to make sure the review-and-delivery-process is met.

morantr commented on tests/test_flows.py:
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_when_not_found: bool = True) -> bool:
+        '''Verify that the logfile for "entity_or_radio" matches "regex",
+           fail if no match when fail_when_not_found is enabled.
```
I would change `fail_when_not_found ` to `fail_on_mistmatch".
Also, please wrap the argument name on the comments with double-quotes to make it clear that it is an argument and not regular text.

orenvor replied:
applied suggestion.

morantr reviewed (approved):
Good PR.
I would split the change on `test_flows` to a second commit and explaining on the commit message that due to the change on the handler location, the `test_flow` will break and will be fixed on the next commit.
But since you already did it like this, it is fine by me to leave it that way.

adam1985d commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -83,7 +83,6 @@
 #include <tlvf/wfa_map/tlvClientInfo.h>
 #include <tlvf/wfa_map/tlvErrorCode.h>
-#include <tlvf/wfa_map/tlvHigherLayerData.h>
 #include <tlvf/wfa_map/tlvMetricReportingPolicy.h>
```
the commit title is still too long (see contribution.md):
you can replace with:
```
agent: move 1905_higher_layer_data handling to son_slave
```

morantr replied:
The commits message is ok. As long there is no `...` at the end, it is ok.

adam1985d reviewed (approved):
I approved but please change the commit title to meet the length limit

VladyslavTupikin commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4198,6 +4201,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+                                                         ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    // Only one backhaul manager (the slave) should return ACK for higher layer data message.
+    if (is_backhaul_manager) {
```
Do we need this check here?

morantr replied:
Yes.

RanRegev commented on tests/test_flows.py:
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
-                  start_line: int = 0) -> bool:
-        '''Verify that the logfile for "entity_or_radio" matches "regex", fail if not.'''
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_on_mistmatch: bool = True) -> bool:
```
I don't understand this change:
changes to `check_log` should be done in a separate commit, with its own rational - as it serves other tests as well.

orenvor replied:
Currently, the `check_log`  function is used as an "assert" statement, 
designed to fail the test when a string was not found in the log.

This change allows testing of the logs without the "assert" part, 
for cases where a missing string is expected or acceptable.

VladyslavTupikin commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -234,6 +234,7 @@ class slave_thread : public beerocks::socket_thread {
     beerocks::message::sWifiChannel channel_selection_select_channel();
     bool handle_multi_ap_policy_config_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
     bool handle_client_association_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_higher_layer_data_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
```suggestion
    bool handle_1905_higher_layer_data_message(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```

orenvor replied:
suggestion applied.

VladyslavTupikin commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4198,6 +4201,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
     return send_cmdu_to_controller(cmdu_tx);
 }
 
+bool slave_thread::handle_1905_higher_layer_data_message(Socket *sd,
```
```suggestion
bool slave_thread::handle_1905_higher_layer_data_message(Socket &sd,
```

orenvor replied:
suggestion applied.

VladyslavTupikin commented on tests/test_flows.py:
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
-                  start_line: int = 0) -> bool:
-        '''Verify that the logfile for "entity_or_radio" matches "regex", fail if not.'''
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_on_mistmatch: bool = True) -> bool:
```
Changes in test_flows should be done in another commit.

orenvor replied:
fixed.

RanRegev commented on tests/test_flows.py:
```diff
@@ -1246,14 +1249,27 @@ def test_higher_layer_data_payload_trigger(self):
+        debug("Confirming higher layer data message was received in one of the agent's radios")
 
-        self.check_log(env.agents[0], r"HIGHER_LAYER_DATA_MESSAGE")
+        received_in_radio0, _, _ = self.check_log(env.agents[0].radios[0],
```
A comment only:
We should stop using check_log to verify correctness.
We can either:
* query the application (best)
* check packets (good)
I leave it as a comment only because this is not a new `check_log`.

VladyslavTupikin commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1509,8 +1509,8 @@ bool master_thread::handle_cmdu_1905_higher_layer_data_message(const std::string
     const auto payload_length = tlvHigherLayerData->payload_length();
-    LOG(DEBUG) << "protocol: " << std::hex << protocol;
-    LOG(DEBUG) << "payload length: " << int(payload_length);
+    LOG(DEBUG) << "Protocol: " << std::hex << protocol;
```
These changes should be done in another commit.

orenvor replied:
fixed.

RanRegev reviewed (approved):
Approving, 
though  I would like to see the changes to `check_log` function introduced in a different PR.

orenvor committed:
agent: move 1905_higher_layer_data handling from backhaul to son_slave

currently the backhaul manager handles 1905_higher_layer_data messages,
while they should be handled by the agent. This PR moves the handling
to son_slave thread as preparation for future unified agent class.

This commit breaks higher_layer_data_payload_trigger test in
test_flows.py. The test will be fixed in the next commit.

Move handle_1905_higher_layer_data_message() from the backhaul manager
to the son_slave thread.

PPM-334

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
test_flows.py: check_log: allow check of log without fail on mismatch

The current implementation of check_log() function is design to fail
the test when a "regex" match was not found in a log-file.

This commit adds an optional "fail_on_mismatch" argument to the
"check_log()" function, to allow testing of logs without failing
the test, in cases where a "regex" mismatch is expected or acceptable.

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
tests: test_higher_layer_data_payload_trigger: ACK from son_slave

update higher_layer_data_payload_trigger test to expect ACK response
from a single son_slave thread instead of the current response from
backhaul manager.

PPM-334

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>
